### PR TITLE
Add the CCM device to be able to generate an encrypted file-system upfront the unikernel

### PR DESCRIPTION
**Pull request by dinosaure**
This PR adds a new device which compose a _block device_ with `mirage-ccm-block` and provide to the `unikernel.ml` an encrypted block-device which can be used transparently with a file-system such as `chamelon`. This PR still is a draft when we are waiting a release of `mirage-ccm-block`.

The name `ccm` is probably not the best and if you have a better idea, I can change it.

This PR is related to:
- https://github.com/ocaml/opam-repository/pull/21122
- https://github.com/sg2342/mirage-block-ccm/issues/27
- and https://github.com/sg2342/mirage-block-ccm/pull/28

But also:
- https://github.com/palainp/mirage-sshfs/issues/11

The idea is to provide something which is much more composable for `mirage-sshfs` and easily switch the file-system used by `mirage-sshfs` by something else than `chamelon` and keep the encryption layer.

### CODE REVIEW

**Comment by palainp** [comment]
I'm of course biased, but I totally approve this PR :)

**Comment by hannesm** [comment]
Another option is to use an empty block device and pass it through mirage-block-ccm? I'd be in favour of having this in the documentation as well (next to using `ccmblock` on an existing block device).

**Comment by reynir** [comment]
Could it be renamed `ccm_block` (or `block_ccm`)? To better reflect that it is ccm+block device

**Comment by dinosaure** [comment]
I improved the documentation to say that it's not needed to encrypt a block device _upstream_. It only matters when your block device contains data (like the `chamelon` format). I'm not sure about how to explain that clearly. I renamed `ccm` to `ccm_block` too.

File: lib/mirage/impl/mirage_impl_block.ml [hunk 1]
```diff
@@ -287,3 +287,26 @@ let chamelon ~program_block_size =
   in
   impl ~packages ~keys ~connect "Kv.Make"
     (block @-> pclock @-> Mirage_impl_kv.rw)
+
+let ccm ?maclen ?nonce_len key =
+  let keys = [ Key.v key ] in
+  let packages =
+    [ package "mirage-block-ccm" ~min:"1.0.2" ~max:"1.1.0"; package "astring" ]
```
[the remark this answers is not in the document]
Comment (hannesm) [reply lib/mirage/impl/mirage_impl_block.ml]: done in e48e9601

**Comment by dinosaure** [comment]
I will try to provide an example on `mirage-skeleton` and we will be ready to merge!

File: lib/mirage/impl/mirage_impl_block.ml [hunk 2]
```diff
@@ -280,10 +280,32 @@ let chamelon ~program_block_size =
     | [ block; _ ] ->
         Fmt.str
           {ocaml|%s.connect ~program_block_size:%a %s
-                 >|= Rresult.R.reword_error (Rresult.R.msgf "%%a" %s.pp_error)
-                 >|= Rresult.R.failwith_error_msg|ocaml}
+                 >|= Result.map_error (Fmt.str "%%a" %s.pp_error)
+                 >|= Result.fold ~ok:(fun x -> x) ~error:failwith|ocaml}
```
Comment (hannesm) [inline lib/mirage/impl/mirage_impl_block.ml:284]:
```suggestion
                 >|= Result.fold ~ok:Fun.id ~error:failwith|ocaml}
```

**Pull request merged**